### DISCUSSION TITLE
How to set the color of TButton's ???

### DISCUSSION THREAD

**Discussion by paule32** [magiblot/tvision#119] [Q&A] [answered]
Hello,
I have the following Code, and I would like to set the Color of TButton, because in the current TWindow, they will be displayed in red, and  NOT in green/black:
```
class TMyButton : public TButton {
public:
    TMyButton(
        const TRect& bounds,
        const char *aTitle,
        ushort aCommand,
        ushort aFlags) :
        TButton(bounds, aTitle, aCommand, aFlags) {}

    void draw() override {
        TDrawBuffer b;
        ushort color;

        // Überprüfe, ob die linke Maustaste gedrückt ist
        if (state & sfActive && state & sfSelected) {
            color = getColor(2);  // Farbindex für den aktiven Zustand
        } else {
            color = getColor(1);  // Farbindex für den inaktiven Zustand
        }

        b.moveChar(0, ' ', color, size.x);
        writeBuf(0, 0, size.x, 1, b);
    }
};

class TMyTvEditor : public TWindow {
public:
    class MyEditorChild: public TFileEditor {
    public:
        MyEditorChild(
            TMyTvEditor* parent,
            const TRect& bounds,
            TScrollBar * hScrollBar,
            TScrollBar * vScrollBar,
            TIndicator * indicator,
            TStringView  filename):
            TFileEditor( bounds,hScrollBar,vScrollBar,indicator,filename),
            owner(parent) { }
        void handleEvent( TEvent &event )
        {
            TFileEditor::handleEvent( event );
            if (event.what == evKeyDown) {
                if (event.keyDown.charScan.charCode == 27)     // #27 - Escape
                {
                    owner->close();
                    clearEvent(event);
                    return;
                }   else
                if (event.keyDown.keyCode == 0x3c00)  // F2
                {
                    clearEvent(event);
                    messageBox("getkey F2",mfInformation|mfOKButton);
                    return;
                }
            }   else
            if (event.what == evCommand) {
                if (event.message.command == 9)      // F1  - Function key
                {
                    clearEvent(event);
                    messageBox("getkey F1",mfInformation|mfOKButton);
                    return;
                }
            }
        }
    private:
        TMyTvEditor * owner;
    };

    ~TMyTvEditor() {
    }
    TMyTvEditor(const TRect& bounds):
        TWindowInit(&TMyTvEditor::initFrame),
        TWindow(bounds, "Mein Fenster", wnNoNumber) {

        hScrollBar = new TScrollBar( TRect( 18, size.y - 1, size.x - 23, size.y ) );
        insert(hScrollBar);

        vScrollBar = new TScrollBar( TRect( size.x - 23, 1, size.x - 22, size.y - 1 ) );
        insert(vScrollBar);

        indicator = new TIndicator( TRect( 2, size.y - 1, 16, size.y) );
        insert(indicator);
        
        editor = new MyEditorChild(this, TRect(1, 1, size.x - 24, size.y - 1), hScrollBar, vScrollBar, indicator, "www.txt");
        insert(editor);

        button1 = new TMyButton(TRect(size.x - 21, 2, size.x - 3, 4), "Schaltfläche 1", cmButton1, bfNormal);
        insert(button1);

        button2 = new TMyButton(TRect(size.x - 21, 5, size.x - 3, 7), "Schaltfläche 2", cmButton2, bfNormal);
        insert(button2);
        
        options |= ofSelectable;
    }

    void handleEvent(TEvent & event ){
        TWindow::handleEvent( event );
        /*
        if (event.what == evMouseDown) {
            if (event.mouse.buttons == mbLeftButton) {
                int mouseX = event.mouse.where.x;
                int mouseY = event.mouse.where.y;
                if ((mouseX == 1) && (mouseY == 1)) {
                    close();
                }   clearEvent(event);
            }
        }   else*/
        if (event.what == evCommand) {
            switch (event.message.command) {
                case cmClose:
                    close();
                    clearEvent(event);
                    break;
                case cmButton1:
                    // Aktion für Schaltfläche 1
                    break;
                case cmButton2:
                    // Aktion für Schaltfläche 2
                    break;
            }
        }
    }

private:
    MyEditorChild * editor;
    TIndicator    * indicator;
    
    TScrollBar    * vScrollBar;
    TScrollBar    * hScrollBar;
    
    TMyButton     * button1;
    TMyButton     * button2;
    
    ushort myButtonColor = 0x2F;
};

```
Here, a screen shoot:
![grafik](https://github.com/magiblot/tvision/assets/3682603/b5f70184-71d6-463c-bfdd-db8925aa10c8)

**Comment by magiblot** [comment]
You need `TMyTvEditor` to inherit from `TDialog` instead of `TWindow`. `TButton` and the other widgets defined in `dialogs.h` are meant to be used in `TDialog` windows. The reason for this is Turbo Vision's intricate palette system.

**Reply by paule32** [reply]
but TDialog's, they aren't modal ?

**Reply by magiblot** [reply]
Not necessarily. That's only when you use the `execView` method. If you just use the `insert` method to add your dialog to the application, it will not be modal. In this sense they behave the same as `TWindow`s.

**Reply by paule32** [reply]
ah, I see. Okay. Now, it is gray, and I don't can see the Indicator - because it is gray/white.
Can I change the Color to someone else ?

**Reply by magiblot** [reply, marked as answer]
Yes, in the constructor of `TMyTvEditor` (the dialog view) add the following line:
```c++
palette = dpBlueDialog;
```

**Reply by paule32** [reply]
Now, the TButtons are not updated - only, when I click on them, they will be display okay.

**Reply by paule32** [reply]
there are screen shoot:
![grafik](https://github.com/magiblot/tvision/assets/3682603/deed132f-b968-4f1f-bad6-ef1313e1dc6f)

**Reply by magiblot** [reply]
Do you still have the custom `TMyButton` class? If so, remove it and use `TButton` directly.

**Comment by paule32** [comment]
Hello,
now, I have remove the TMyButton class - it is not necessary. A simple TButton * button = new ... is enough.
I was "override" the "draw" member, and as such, the TButton paint it self - all not needed.
For those of interest, I share my Code:
```
class TMyTvEditor : public TDialog {
public:
    class MyEditorChild: public TFileEditor {
    public:
        MyEditorChild(
            TMyTvEditor* parent,
            const TRect& bounds,
            TScrollBar * hScrollBar,
            TScrollBar * vScrollBar,
            TIndicator * indicator,
            TStringView  filename):
            TFileEditor( bounds,hScrollBar,vScrollBar,indicator,filename),
            owner(parent) { }
        void handleEvent( TEvent &event )
        {
            TFileEditor::handleEvent( event );
            if (event.what == evKeyDown) {
                if (event.keyDown.charScan.charCode == 27)     // #27 - Escape
                {
                    owner->close();
                    clearEvent(event);
                    return;
                }   else
                if (event.keyDown.keyCode == 0x3c00)  // F2
                {
                    clearEvent(event);
                    messageBox("getkey F2",mfInformation|mfOKButton);
                    return;
                }
            }   else
            if (event.what == evCommand) {
                if (event.message.command == 9)      // F1  - Function key
                {
                    clearEvent(event);
                    messageBox("getkey F1",mfInformation|mfOKButton);
                    return;
                }
            }
        }
    private:
        TMyTvEditor * owner;
    };

    ~TMyTvEditor() {
    }
    TMyTvEditor(const TRect& bounds):
        TWindowInit(&TMyTvEditor::initFrame),
        TDialog(bounds, "Mein Fenster") {
        palette = dpBlueDialog;

        hScrollBar = new TScrollBar( TRect( 18, size.y - 1, size.x - 23, size.y ) );
        insert(hScrollBar);

        vScrollBar = new TScrollBar( TRect( size.x - 23, 1, size.x - 22, size.y - 1 ) );
        insert(vScrollBar);

        indicator = new TIndicator( TRect( 2, size.y - 1, 16, size.y) );
        insert(indicator);
        
        editor = new MyEditorChild(this, TRect(1, 1, size.x - 24, size.y - 1), hScrollBar, vScrollBar, indicator, "www.txt");
        insert(editor);

        button1 = new TButton(TRect(size.x - 21, 2, size.x - 3, 4), "Schaltfläche 1", cmButton1, bfNormal);
        insert(button1);

        button2 = new TButton(TRect(size.x - 21, 5, size.x - 3, 7), "Schaltfläche 2", cmButton2, bfNormal);
        insert(button2);
        
        options |= ofSelectable;
        
        drawView();
    }

    void handleEvent(TEvent & event ){
        TWindow::handleEvent( event );
        if (event.what == evCommand) {
            switch (event.message.command) {
                case cmClose:
                    close();
                    clearEvent(event);
                    break;
                case cmButton1:
                    // Aktion für Schaltfläche 1
                    break;
                case cmButton2:
                    // Aktion für Schaltfläche 2
                    break;
            }
        }
    }

private:
    MyEditorChild * editor;
    TIndicator    * indicator;
    
    TScrollBar    * vScrollBar;
    TScrollBar    * hScrollBar;
    
    TButton       * button1;
    TButton       * button2;
    
    ushort myButtonColor = 0x2F;
};
```
And here is a  screen shoot:
![grafik](https://github.com/magiblot/tvision/assets/3682603/1f96466f-8228-4c90-915c-ba224582fb2b)

Thanks a lot to magiblot !

**Comment by magiblot** [comment]
> I prefer, to color the Border, and Indicator as gray/black - if possible ?

Yes, but since Turbo Vision does not come with such a palette, you will have to set the colors manually. The simplest way is to override the `mapColor` method of `TMyTvEditor` (the dialog view). From there you will be able to control the color of all the views inserted into the dialog. Here are some insights (information taken from [Extended color support](https://github.com/magiblot/tvision#extended-color-support)):

> `mapColor` looks up a single color attribute in the view's palette, given an index into the palette. Remember that the palette indices for each view class can be found in the Turbo Vision headers. For example, `<tvision/views.h>` says the following about `TScrollBar`:
> 
> ```
> /* ---------------------------------------------------------------------- */
> /*      class TScrollBar                                                  */
> /*                                                                        */
> /*      Palette layout                                                    */
> /*        1 = Page areas                                                  */
> /*        2 = Arrows                                                      */
> /*        3 = Indicator                                                   */
> /* ---------------------------------------------------------------------- */
> ```

> As an API extension, the mapColor method has been made virtual. This makes it possible to override Turbo Vision's hierarchical palette system with a custom solution without having to rewrite the draw() method.
> 
> So, in general, there are three ways to use extended colors in views:
> 
> 3. By returning extended color attributes from an overriden `mapColor` method:
> 
> ```c++
> // The 'TMyScrollBar' class inherits from 'TScrollBar' and overrides 'TView::mapColor'.
> TColorAttr TMyScrollBar::mapColor(uchar index) noexcept
> {
>     // In this example the values are hardcoded,
>     // but they could be stored elsewhere if desired.
>     switch (index)
>     {
>         case 1:     return {0x492983, 0x826124}; // Page areas.
>         case 2:     return {0x438939, 0x091297}; // Arrows.
>         case 3:     return {0x123783, 0x329812}; // Indicator.
>         default:    return errorAttr;
>     }
> }
> ```

**Reply by paule32** [reply]
Next mistake:
when I use the code that I had show you, then I have a next problem:
when I open a new/next TDialog/TWindow, and switch between the Windows, I lost display Informations.
So, the h/v TScrollBar's will disappear when I focus a new window, but come back when the background window becomes focus.
There is a screen shoot:
![grafik](https://github.com/magiblot/tvision/assets/3682603/94cb2b0b-cf4e-4d26-8151-acab8258882d)
